contracts/staking/stakeManager/ValidatorAuction.sol

```diff
@@ -0,0 +1,126 @@
+import {GovernanceLockable} from "../../common/mixin/GovernanceLockable.sol";
+import {RootChainable} from "../../common/mixin/RootChainable.sol";
+import {StakingInfo} from "../StakingInfo.sol";
+import {StakingNFT} from "./StakingNFT.sol";
+import {StakeManager} from "./StakeManager.sol";
+import {StakeManagerStorage} from "./StakeManagerStorage.sol";
+import {Math} from "openzeppelin-solidity/contracts/math/Math.sol";
+
+contract ValidatorAuction is StakeManagerStorage {
+    using SafeMath for uint256;
+
+    constructor() public GovernanceLockable(address(0x0)) {}
+
+    function startAuction(
+        uint256 validatorId,
+        uint256 amount,
+        bool _acceptDelegation,
+        bytes calldata _signerPubkey
+    ) external {
+        uint256 currentValidatorAmount = validators[validatorId].amount;
+
+        require(
+            validators[validatorId].deactivationEpoch == 0 && currentValidatorAmount != 0,
+            "Invalid validator for an auction"
+        );
+        uint256 senderValidatorId = signerToValidator[msg.sender];
+        // make sure that signer wasn't used already
+        require(
+            NFTContract.balanceOf(msg.sender) == 0 && // existing validators can't bid
+                senderValidatorId != INCORRECT_VALIDATOR_ID,
+            "Already used address"
+        );
+
+        uint256 _currentEpoch = currentEpoch;
+        uint256 _replacementCoolDown = replacementCoolDown;
+        // when dynasty period is updated validators are in cooldown period
+        require(_replacementCoolDown == 0 || _replacementCoolDown <= _currentEpoch, "Cooldown period");
+        // (auctionPeriod--dynasty)--(auctionPeriod--dynasty)--(auctionPeriod--dynasty)
+        // if it's auctionPeriod then will get residue smaller then auctionPeriod
+        // from (CurrentPeriod of validator )%(auctionPeriod--dynasty)
+        // make sure that its `auctionPeriod` window
+        // dynasty = 30, auctionPeriod = 7, activationEpoch = 1, currentEpoch = 39
+        // residue 1 = (39-1)% (7+30), if residue <= auctionPeriod it's `auctionPeriod`
+
+        require(
+            (_currentEpoch.sub(validators[validatorId].activationEpoch) % dynasty.add(auctionPeriod)) < auctionPeriod,
+            "Invalid auction period"
+        );
+
+        uint256 perceivedStake = currentValidatorAmount;
+        perceivedStake = perceivedStake.add(validators[validatorId].delegatedAmount);
+
+        Auction storage auction = validatorAuction[validatorId];
+        uint256 currentAuctionAmount = auction.amount;
+
+        perceivedStake = Math.max(perceivedStake, currentAuctionAmount);
+
+        require(perceivedStake < amount, "Must bid higher");
+        require(token.transferFrom(msg.sender, address(this), amount), "Transfer failed");
+
+        //replace prev auction
+        if (currentAuctionAmount != 0) {
+            require(token.transfer(auction.user, currentAuctionAmount), "Bid return failed");
```

Comment:
Nothing i guess, we don't really take into account such edge cases.

Reply:
But if recursively repeated than it can drain the funds in the contract. Ideally, if the bid is smaller than the gas then the bidder should not be refunded. This is not related to the pr i just thought, so pointing it out.